File: source/python/66. Plus One.py
class Solution:
    def plusOne(self, digits: list[int]) -> list[int]:
        carry = 1

        for i in range(len(digits) - 1, -1, -1):
            digits[i] += carry
            carry = digits[i] // 10
            digits[i] %= 10

        if carry:
            digits.insert(0, carry)

        return digits


def test():
    s = Solution()
    assert s.plusOne([1, 2, 3]) == [1, 2, 4]
    assert s.plusOne([4, 3, 2, 1]) == [4, 3, 2, 2]
    assert s.plusOne([0]) == [1]
    assert s.plusOne([9]) == [1, 0]
    assert s.plusOne([9, 9]) == [1, 0, 0]
    assert s.plusOne([9, 9, 9]) == [1, 0, 0, 0]
    assert s.plusOne([1, 9, 9]) == [2, 0, 0]

    print("All tests passed!")


if __name__ == "__main__":
    test()

File: source/cpp/66. Plus One.cpp
#include <vector>


class Solution {
public:
    std::vector<int> plusOne(std::vector<int>& digits) {
        auto carry = 1;
        for (int i = digits.size() - 1; i > -1; --i) {
            digits[i] += carry;
            
            if (digits[i] > 9)
                digits[i] %= 10;
            else 
                return digits;
        }

        digits.insert(digits.begin(), 1);
        return digits;
    }
};